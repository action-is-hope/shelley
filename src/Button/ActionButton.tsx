import React, { forwardRef } from "react";
import { st, classes } from "./actionButton.st.css";
import { ButtonBase, ButtonProps } from "./";

interface ActionButtonProps
  extends Omit<ButtonProps, "elementType" | "href" | "isCta"> {
  isQuiet?: boolean;
}

function ActionButton(
  props: ActionButtonProps,
  ref: React.Ref<HTMLButtonElement>
) {
  const { className: classNameProp, isQuiet, ...rest } = props;

  return (
    <ButtonBase
      className={st(
        classes.root,
        {
          isQuiet,
        },
        classNameProp
      )}
      ref={ref}
      {...rest}
    />
  );
}
ActionButton.displayName = "ActionButton";

const _ActionButton = forwardRef(ActionButton);
export { _ActionButton as ActionButton };
