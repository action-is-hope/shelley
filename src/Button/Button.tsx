"use client";
import React, {
  Ref,
  ReactElement,
  useRef,
  forwardRef,
  ElementType,
} from "react";
import { useButton } from "react-aria";
import type { AriaButtonProps } from "@react-types/button";
import type {
  Accent,
  AlignPos,
  Volume,
  ButtonVariants,
} from "../typings/shared-types";
import type { MergeElementProps } from "../typings/utils";
import { useFocusRing } from "react-aria";
import { mergeRefs, mergeProps } from "@react-aria/utils";
import { st, classes } from "./button.st.css";

/**
 * Leveraging Adobes 'useButton' to enable use with their hooks.
 * N.B: Omitting 'elementType' as preference is to use existing 'as'.
 *
 * Adobe docs: https://react-spectrum.adobe.com/react-aria/useButton.html
 */

export interface ButtonCustomProps
  extends Omit<AriaButtonProps, "elementType"> {
  /** Define an Icon node, postion via #iconPos. */
  icon?: React.ReactNode;
  /** The position of the icon relative to the label. */
  iconPos?: AlignPos;
  /** Tone index, defines the color palette. */
  tone?: Accent;
  /** Variant index, defines the 'look'. */
  variant?: ButtonVariants;
  /** Defines how 'loud' the Button should be in term of it's size. */
  vol?: Volume;
  /** Applies width of 100%. */
  fullWidth?: boolean;
}

export type ButtonProps<P extends React.ElementType = "a" | "button" | any> = {
  /** Custom element to render such as an anchor "a" or a router "Link" component. */
  as?: P;
  // Dynamically apply element props types based on the input (P).
} & MergeElementProps<P, Omit<ButtonCustomProps, "as">>;

function Button<P extends React.ElementType = "button">(
  {
    as: As,
    children,
    className: classNameProp,
    icon,
    iconPos,
    fullWidth = false,
    tone = 1,
    variant = "primary",
    vol = 3,
    // Pull off known inputs for @react-aria -> useButton minus 'elementType'
    excludeFromTabOrder,
    isDisabled,
    onPress,
    onPressStart,
    onPressEnd,
    onPressChange,
    onPressUp,
    // autoFocus,
    onFocus,
    onBlur,
    onFocusChange,
    onKeyDown,
    onKeyUp,
    href,
    target,
    rel,
    ariaExpanded,
    ariaHaspopup,
    ariaControls,
    ariaPressed,
    type,
    id,
    ariaLabel,
    ariaLabelledby,
    ariaDescribedby,
    ariaDetails,
    ...rest
  }: ButtonProps<P>,
  ref: React.Ref<HTMLElement>
) {
  const localRef = useRef(null);

  const { buttonProps, isPressed } = useButton(
    {
      excludeFromTabOrder,
      isDisabled,
      onPress,
      onPressStart,
      onPressEnd,
      onPressChange,
      onPressUp,
      // autoFocus,
      onFocus,
      onBlur,
      onFocusChange,
      onKeyDown,
      onKeyUp,
      href,
      target,
      rel,
      "aria-expanded": ariaExpanded,
      "aria-haspopup": ariaHaspopup,
      "aria-controls": ariaControls,
      "aria-pressed": ariaPressed,
      type,
      id,
      "aria-label": ariaLabel,
      "aria-labelledby": ariaLabelledby,
      "aria-describedby": ariaDescribedby,
      "aria-details": ariaDetails,
      // Map 'as' to elementType for adobe-aria...
      elementType: (As as React.JSXElementConstructor<HTMLElement>) || "button",
    },
    localRef as React.RefObject<HTMLElement>
  );
  const { isFocusVisible, focusProps } = useFocusRing();

  const className = st(
    classes.root,
    {
      iconPos: icon ? iconPos : undefined,
      fullWidth,
      tone: tone !== false ? tone : undefined,
      variant: variant !== false ? variant : undefined,
      vol: vol !== false ? vol : undefined,
      isFocusVisible,
      isPressed,
      isDisabled,
    },
    classNameProp
  );

  const internal = (
    <>
      {icon && (
        <>
          {icon}
          {children && <span className={classes.divider}></span>}
        </>
      )}
      {children && <span className={classes.inner}>{children}</span>}
    </>
  );
  return React.createElement(
    As || "button",
    {
      ref: mergeRefs(localRef, ref),
      ...mergeProps(buttonProps, focusProps),
      className,
      ...rest,
    },
    internal
  );
}

/**
 * Buttons allow users to perform an action.
 */
// forwardRef doesn't support generic parameters -> cast to the correct type.
// https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref
const _Button = forwardRef(Button) as <P extends ElementType>(
  props: ButtonProps<P> & { ref?: Ref<HTMLElement> }
) => ReactElement;
/** required - see ButtonGroup */
_Button.toString = () => "ShelleyButton";
export { _Button as Button };
