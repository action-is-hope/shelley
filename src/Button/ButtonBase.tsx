"use client";
import React, { useRef, forwardRef } from "react";
import { useButton } from "react-aria";
import type { AriaButtonProps } from "@react-types/button";
import type {
  AlignPos,
  Volume,
  ExtendedToneVariants,
  ExtendedButtonVariants,
} from "../typings/shared-types";
import type { MergeElementProps } from "../typings/utils";
import { useFocusRing } from "react-aria";
import { mergeRefs, mergeProps } from "@react-aria/utils";
import {
  st as stButtonBase,
  classes as classesButtonBase,
} from "./buttonBase.st.css";

export interface ButtonProps<P = "button", V = "", T = "">
  extends Omit<AriaButtonProps, "elementType" | "href"> {
  /** Custom `className` for overriding styles. */
  className?: string;
  /** Custom element to render such as an anchor "a" or a router "Link" component. */
  elementType?: P;
  /** Define an Icon node, postion via #iconPos. */
  icon?: React.ReactNode;
  /** The position of the icon relative to the label. */
  iconPos?: AlignPos;
  /** Tone index, defines the color palette. */
  tone?: ExtendedToneVariants<T>;
  /** Variant defines the 'look'. */
  variant?: ExtendedButtonVariants<V>;
  /** Defines how 'loud' the Button should be in term of it's size. */
  vol?: Volume;
  /** Applies width of 100%. */
  fullWidth?: boolean;
  /** Is the button a Call To Action */
  isCta?: boolean;
}

function ButtonBase<
  V extends string,
  T extends string,
  P extends React.ElementType = "button"
>(
  props: MergeElementProps<P, ButtonProps<P, V, T>>,
  ref: React.Ref<HTMLElement>
) {
  const {
    elementType: Element = "button",
    children,
    className: classNameProp,
    icon,
    iconPos,
    fullWidth = false,
    tone,
    variant,
    vol = 3,
    isDisabled,
    ...rest
  } = props;

  const internalRef = useRef(null);

  const { buttonProps, isPressed } = useButton(
    props,
    internalRef as React.RefObject<HTMLElement>
  );
  const { isFocusVisible, focusProps } = useFocusRing();

  const className = stButtonBase(
    classesButtonBase.root,
    {
      iconPos: icon ? iconPos : undefined,
      variant: variant || undefined,
      tone: tone || undefined,
      vol: vol || undefined,
      isFocusVisible,
      isPressed,
      isDisabled,
      fullWidth,
    },
    classNameProp
  );

  const internal = (
    <>
      {children && <span className={classesButtonBase.inner}>{children}</span>}
      {icon && (
        <>
          {children && <span className={classesButtonBase.divider}></span>}{" "}
          {icon}
        </>
      )}
    </>
  );
  return React.createElement(
    Element,
    {
      ref: mergeRefs(internalRef, ref),
      ...mergeProps(buttonProps, focusProps),
      className,
      ...rest,
    },
    internal
  );
}
ButtonBase.displayName = "Button";

/**
 * Button component capable of adapting to various element types (e.g., "button", "a", or a router "Link").
 * Utilizes `react-aria` for accessibility features. see {@link http://example.com|Example Website}
 */
const _ButtonBase = forwardRef(ButtonBase);
_ButtonBase.toString = () => "ShelleyButton";
export { _ButtonBase as ButtonBase };
