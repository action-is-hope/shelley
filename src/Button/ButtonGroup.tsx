"use client";
import React, {
  Children,
  cloneElement,
  forwardRef,
  isValidElement,
} from "react";
import { mergeProps } from "react-aria";
import type { ButtonProps } from "./";
import type {
  ExtendedToneVariants,
  ExtendedButtonVariants,
  Volume,
} from "../typings/shared-types";
import { st, classes } from "./buttonGroup.st.css";

export interface ButtonGroupProps<V = "", T = "">
  extends React.HTMLAttributes<HTMLDivElement> {
  /** Adds a class to each button. */
  buttonClassName?: string;
  /** Disables all the buttons. */
  isDisabled?: boolean;
  /** Tone index. */
  tone?: ExtendedToneVariants<T>;
  /** Variant index. */
  variant?: ExtendedButtonVariants<V>;
  /** Changes the volume of the buttons. */
  vol?: Volume;
  /** Orient around vertical or horizontal. */
  orientation?: "vertical" | "horizontal";
  /** Use a split button style. */
  splitButton?: boolean;
}

function ButtonGroup<V extends string, T extends string>(
  props: ButtonGroupProps<T, V>,
  ref: React.Ref<HTMLDivElement>
) {
  const {
    buttonClassName,
    children,
    className: classNameProp,
    isDisabled,
    tone = "lead",
    variant = "quiet",
    orientation = "horizontal",
    vol = 3,
    splitButton,
    ...rest
  } = props;
  return (
    <div
      className={st(
        classes.root,
        { orientation, splitButton, isDisabled },
        classNameProp
      )}
      {...rest}
      ref={ref}
    >
      {Children.map(children, (child) => {
        if (!isValidElement(child)) {
          return null;
        }
        // Check if Child is a Shelley Button
        return child.type?.toString() === "ShelleyButton"
          ? cloneElement(child, {
              ...mergeProps(
                { isDisabled, tone, variant, vol },
                { ...(child.props as ButtonProps) }
              ),
            })
          : child;
      })}
    </div>
  );
}
ButtonGroup.displayName = "ButtonGroup";

/**
 * ButtonGroup can be used to group related buttons in various orientations.
 */
const _ButtonGroup = forwardRef(ButtonGroup);
export { _ButtonGroup as ButtonGroup };
