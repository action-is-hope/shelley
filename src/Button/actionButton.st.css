@st-namespace "ActionButton";
@st-import ButtonBase, [
  --bg-color,
  --border-color,
  --text-color,
  --border-width,
  --border-radius] from "./buttonBase.st.css";

.root {
  -st-extends: ButtonBase;
  -st-states: isQuiet;
}
