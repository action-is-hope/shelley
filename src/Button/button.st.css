@namespace "Button";

.root {
  -st-states:
    iconPos(enum(top, end, bottom, start)),
    fullWidth,
    tone(number),
    variant(string),
    vol(number),
    isFocusVisible,
    isPressed,
    isDisabled;
}

.inner {}

/* The divider which is rendered with an icon only. */
.divider {}

.root,
.root:visited {
  /* = Local vars. */
  --bg-color: transparent;
  --border-color: transparent;
  --text-color: inherit;
  --border-width: 1px;
  --border-radius: 4px;
  display: inline-flex;
  justify-content: space-between;
  /* = Explicitly set font-family else it will default to system style. */
  font-family: inherit;
  /* = Vertical content alignment - kicks in when heights are 'known'. */
  align-items: center;
  /* = Persist border to maintain alignment should you be using borders as 
    part of additonal visual states. */
  border-style: solid;
  border-width: var(--border-width);
  border-radius: var(--border-radius);
  border-color: var(--border-color);
  color: var(--text-color);
  background-color: var(--bg-color);
  /* = Horizontal alignment. */
  justify-content: center;
  /* = Safari spacing issue with button groups MacOs and IOS. */
  margin: 0;
  text-decoration: none;
}

/*==== Icon Positions ====*/

/* top */
.root:iconPos(top) {
  flex-direction: column;
}

/* end / right */
.root:iconPos(end) {
  flex-direction: row;
}

/* bottom */
.root:iconPos(bottom) {
  flex-direction: column-reverse;
}

/* start / left */
.root:iconPos(start) {
  flex-direction: row-reverse;
}

.root:fullWidth {
  width: 100%;
}
