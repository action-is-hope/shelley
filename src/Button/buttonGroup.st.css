@st-namespace "btnGroup";
@st-import Button from "../Button/button.st.css";

.root {
  -st-states:
    orientation(enum(horizontal, vertical)),
    fullWidth,
    splitButton,
    isDisabled;
}
