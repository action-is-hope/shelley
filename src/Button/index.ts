export * from "./ButtonBase";
export * from "./Button";
export * from "./ActionButton";
export * from "./IconButton";
