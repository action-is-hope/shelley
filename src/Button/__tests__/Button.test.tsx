import React from "react";
import { Button } from "../Button";
import renderer from "react-test-renderer";
import { Icon } from "../../Icon/Icon";
import { Router, Link } from "react-router-dom";
import { createMemoryHistory } from "history";

const buttonRef = React.createRef<HTMLButtonElement>();

const buttonRefAnchor = React.createRef<HTMLAnchorElement>();
const btnText = "Button text";
const iconPath = <path d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"></path>;

describe("Button", () => {
  it("renders as a basic button with custom #className", () => {
    const tree = renderer
      .create(
        <Button
          data-testid="button-data-testid"
          className="for-which-nobody-can-deny"
          ref={buttonRef}
        >
          {btnText}
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with stylable state prop classes, #tone, #variant, #vol and #fullwidth", () => {
    const tree = renderer
      .create(
        <Button
          data-testid="button-data-testid"
          ref={buttonRef}
          tone={6}
          variant="primary"
          vol={6}
          fullWidth
        >
          {btnText}
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders as a basic anchor link via #href prop", () => {
    const tree = renderer
      .create(
        <Button
          data-testid="button-data-testid"
          href="http://shelley.earth"
          as="a"
        >
          {btnText}
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders as third party (react-router) anchor provided via #as and #to props", () => {
    const memHistory = createMemoryHistory();
    const tree = renderer
      .create(
        <Router history={memHistory}>
          <Button
            as={Link}
            data-testid="button-data-testid"
            to="internal-link"
            ref={buttonRefAnchor}
          >
            {btnText}
          </Button>
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("includes #icon via prop positioned via #iconPos divider and children", () => {
    const tree = renderer
      .create(
        <Button
          data-testid="button-data-testid"
          ref={buttonRef}
          icon={
            <Icon
              // https://github.com/svg/svgo/issues/1011 Still best practice for buttons?
              focusable="false"
            >
              {iconPath}
            </Icon>
          }
          iconPos="start"
        >
          {btnText}
        </Button>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("includes #icon, no children, visually hidden label text provided via Icon alt prop", () => {
    const tree = renderer
      .create(
        <Button
          data-testid="button-data-testid"
          ref={buttonRef}
          variant="secondary"
          aria-label={btnText}
          icon={
            <Icon
              // https://github.com/svg/svgo/issues/1011 Still best practice?
              focusable="false"
            >
              {iconPath}
            </Icon>
          }
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
