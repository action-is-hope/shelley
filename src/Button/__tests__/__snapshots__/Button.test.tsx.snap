// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button includes #icon via prop positioned via #iconPos divider and children 1`] = `
<button
  className="ButtonBase__Test__root ButtonBase__Test---iconPos-5-start ButtonBase__Test---variant-7-primary ButtonBase__Test---tone-4-lead ButtonBase__Test---vol-1-3 Button__Test__root"
  data-id="button-data-id"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
  <span
    className="ButtonBase__Test__divider"
  />
   
  <svg
    aria-hidden={true}
    className="Icon__Test__root"
    focusable="false"
    viewBox="0 0 16 16"
  >
    <path
      d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"
    />
  </svg>
</button>
`;

exports[`Button includes #icon, no children, visually hidden label text provided via Icon alt prop 1`] = `
<button
  aria-label="Button text"
  className="ButtonBase__Test__root ButtonBase__Test---variant-9-secondary ButtonBase__Test---tone-4-lead ButtonBase__Test---vol-1-3 Button__Test__root"
  data-id="button-data-id"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
   
  <svg
    aria-hidden={true}
    className="Icon__Test__root"
    focusable="false"
    viewBox="0 0 16 16"
  >
    <path
      d="M14 7h-5v-5h-2v5h-5v2h5v5h2v-5h5v-2z"
    />
  </svg>
</button>
`;

exports[`Button isLoading renders ProgressCircle 1`] = `
<button
  className="ButtonBase__Test__root ButtonBase__Test---variant-7-primary ButtonBase__Test---tone-4-lead ButtonBase__Test---vol-1-3 ButtonBase__Test--isLoading Button__Test__root"
  data-id="button-data-id"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
  <span
    className="ButtonBase__Test__divider"
  />
   
  <div
    aria-label="Loading"
    aria-valuemax={100}
    aria-valuemin={0}
    className="ProgressCircle__Test__root ProgressCircle__Test--isIndeterminate ProgressCircle__Test---size-5-small ProgressCircle__Test---variant-14-overBackground ButtonBase__Test__loader"
    data-id="button-data-id--progressCircle"
    id="react-aria-3"
    role="progressbar"
  >
    <div
      className="ProgressCircle__Test__track"
    />
    <div
      className="ProgressCircle__Test__fills"
    >
      <div
        className="ProgressCircle__Test__fillMask1"
      >
        <div
          className="ProgressCircle__Test__fillSubMask1"
          data-testid="fillSubMask1"
          style={{}}
        >
          <div
            className="ProgressCircle__Test__fill"
          />
        </div>
      </div>
      <div
        className="ProgressCircle__Test__fillMask2"
      >
        <div
          className="ProgressCircle__Test__fillSubMask2"
          data-testid="fillSubMask2"
          style={{}}
        >
          <div
            className="ProgressCircle__Test__fill"
          />
        </div>
      </div>
    </div>
  </div>
</button>
`;

exports[`Button isLoading renders ProgressCircle instead of icon 1`] = `
<button
  className="ButtonBase__Test__root ButtonBase__Test---iconPos-5-start ButtonBase__Test---variant-7-primary ButtonBase__Test---tone-4-lead ButtonBase__Test---vol-1-3 ButtonBase__Test--isLoading Button__Test__root"
  data-id="button-data-id"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
  <span
    className="ButtonBase__Test__divider"
  />
   
  <div
    aria-label="Loading"
    aria-valuemax={100}
    aria-valuemin={0}
    className="ProgressCircle__Test__root ProgressCircle__Test--isIndeterminate ProgressCircle__Test---size-5-small ProgressCircle__Test---variant-14-overBackground ButtonBase__Test__loader"
    data-id="button-data-id--progressCircle"
    id="react-aria-1"
    role="progressbar"
  >
    <div
      className="ProgressCircle__Test__track"
    />
    <div
      className="ProgressCircle__Test__fills"
    >
      <div
        className="ProgressCircle__Test__fillMask1"
      >
        <div
          className="ProgressCircle__Test__fillSubMask1"
          data-testid="fillSubMask1"
          style={{}}
        >
          <div
            className="ProgressCircle__Test__fill"
          />
        </div>
      </div>
      <div
        className="ProgressCircle__Test__fillMask2"
      >
        <div
          className="ProgressCircle__Test__fillSubMask2"
          data-testid="fillSubMask2"
          style={{}}
        >
          <div
            className="ProgressCircle__Test__fill"
          />
        </div>
      </div>
    </div>
  </div>
</button>
`;

exports[`Button renders as a basic anchor link via #href prop 1`] = `
<a
  className="ButtonBase__Test__root ButtonBase__Test---vol-1-3 button-class"
  data-id="button-data-id"
  href="http://shelley.earth"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
</a>
`;

exports[`Button renders as a basic button with custom #className 1`] = `
<button
  className="ButtonBase__Test__root ButtonBase__Test---variant-7-primary ButtonBase__Test---tone-4-lead ButtonBase__Test---vol-1-3 Button__Test__root button-class"
  data-id="button-data-id"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
</button>
`;

exports[`Button renders as third party (react-router) anchor provided via #elementType and #to props 1`] = `
<a
  className="ButtonBase__Test__root ButtonBase__Test---variant-7-primary ButtonBase__Test---tone-4-lead ButtonBase__Test---vol-1-3 Button__Test__root"
  data-id="button-data-id"
  href="/internal-link"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
</a>
`;

exports[`Button renders with stylable state prop classes, #tone, #variant, #vol and #fullwidth 1`] = `
<button
  className="ButtonBase__Test__root ButtonBase__Test---variant-7-primary ButtonBase__Test---tone-5-alert ButtonBase__Test---vol-1-6 ButtonBase__Test--fullWidth Button__Test__root"
  data-id="button-data-id"
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <span
    className="ButtonBase__Test__inner"
  >
    Button text
  </span>
</button>
`;
